base/promotion.jl

```diff
@@ -5,61 +5,67 @@
 """
     typejoin(T, S)
 
-Compute a type that contains both `T` and `S`.
+
+Return the closest common ancestor of `T` and `S`, i.e. the narrowest type from which
+they both inherit.
 """
 typejoin() = (@_pure_meta; Bottom)
 typejoin(@nospecialize(t)) = (@_pure_meta; t)
 typejoin(@nospecialize(t), ts...) = (@_pure_meta; typejoin(t, typejoin(ts...)))
-function typejoin(@nospecialize(a), @nospecialize(b))
+typejoin(@nospecialize(a), @nospecialize(b)) = (@_pure_meta; join_types(a, b, typejoin))
+
+function join_types(@nospecialize(a), @nospecialize(b), f::Function)
     @_pure_meta
```

Comment:
not pure anymore (due to the abstractly typed `f::Function` parameter)

Reply:
Ah, I always get this wrong... Fixed.

Reply:
Don't we need this function to be pure though?

Reply:
I have no idea. We could use a loop inside `@eval` to define two separate functions which would be pure, without duplicating the code.